executor/update.go

```diff
@@ -241,14 +273,21 @@ func (e *UpdateExec) fastComposeNewRow(rowIdx int, oldRow []types.Datum, cols []
 }
 
 func (e *UpdateExec) composeNewRow(rowIdx int, oldRow []types.Datum, cols []*table.Column) ([]types.Datum, error) {
+	oldRowBuffer := chunk.MutRowFromDatums(oldRow)
 	newRowData := types.CloneRow(oldRow)
 	e.evalBuffer.SetDatums(newRowData...)
-	for _, assign := range e.OrderedList {
+	for i, assign := range e.OrderedList {
 		handleIdx, handleFound := e.tblColPosInfos.FindHandle(assign.Col.Index)
 		if handleFound && e.canNotUpdate(oldRow[handleIdx]) {
 			continue
 		}
-		val, err := assign.Expr.Eval(e.evalBuffer.ToRow())
+		var val types.Datum
+		var err error
+		if i < e.virtualAssignmentsOffset {
+			val, err = assign.Expr.Eval(oldRowBuffer.ToRow())
+		} else {
+			val, err = assign.Expr.Eval(e.evalBuffer.ToRow())
```

Comment:
how about add a swap + generated column test case~?

Reply:
Thanks. I've just added some test cases and found bugs in generated columns of table with multiple aliases. I will fix it soon.